New map generator

berekuk opened the pull request:
# Summary

<s>This is a draft, but I'm at the stage where I want to get some feedback on whether the team is ok with this approach.</s>

`mettagrid.map.mapgen.MapGen` class provides a new way to construct maps, based on scenes that can include other scenes recursively, and you select children areas to do that in bulk, all through YAML configs.

Example: "render a grid of 2x3 rooms; then add two mazes in two random rooms; add 1 agent in each room; add 1 mine and 2 generators in each room; add this handcoded ASCII map in a room that's not a maze".

`mapgen_demo.yaml` config, that can be viewed with `python -m tests.show_map game/map_builder=mapgen_demo game.num_agents=6`, should serve as a detailed summary of current capabilities.

<img width="740" alt="Screenshot 2025-04-07 at 20 23 29" src="https://github.com/user-attachments/assets/d8341c21-5f19-4d98-a3e5-0d3ed17b3488" />


This PR includes the commits from #68, because I needed the `tests.show_map` script for development.

# Notes

## `configs/game/map_builder/` refactoring

I wanted to make configs reusable, so I've moved parts of old `configs/*.yaml` files to separate files.

This caused two problems:
1) If this PR is merged, the main Metta repo will break - it symlinks to only some configs, which I didn't realize initially.
2) In the current design, `game.map_builder` is tightly coupled with `game.num_agents`.

I'd argue that maybe `game.num_agents` shouldn't exist at all — the env constructor could export `map_agents` by counting the agents that are present on the map. But this is not a blocker for this PR.

For symlinks, I can revert the refactorings for older configs. Another thing that could be done is to `ln -s deps/mettagrid/configs configs/env/mettagrid`, and move extra mettagrid configs in Metta repo under `configs/env/mettagrid-extra`. 

## Demos for old -> new system

`configs/game/map_builder/` includes separate `maze.yaml` and `mapgen_maze.yaml` and similar pairs, to check that the new system supports all previous maps.

(I haven't converted cognitive_evals maps yet.)

The maps are not 100% exactly identical, but they're very close.

## Future plans

On my current roadmap:
- BSP tree maze generation (rougelike), since it's the most straightforward way to demo the new capabilities in a less boring environment than separate isolated rooms
  - I'm reimplementing this from scratch, btw, because the full scene implementation needs to export *areas* (rooms and corridors), so I couldn't use [urizen](https://github.com/vurmux/urizen)
- WaveFunctionCollapse integration
- Non-rectangular areas (probably with something like... rectangular area + a bitmask?)
  - Rationale/example: imagine that BSP scene exports "room" areas and "corridor" areas; then we want to populate corridors with items. If the corridor has turns then the rectangular area that covers the corridor could cover a corner of some unrelated room.
- Other scenes, such as:
  - "fill this area with walls of this shape, but make checks that it's still fully connected"
  - Layout scenes that allow you to target sub-areas like "left side of the map", or "top-left corner with 20% width/height"
- Built-in config flag for rotating scenes

berekuk commented:
BSP (roguelike) map generator:

<img width="1392" alt="Screenshot 2025-04-07 at 22 07 44" src="https://github.com/user-attachments/assets/2b357c94-5446-4452-b13d-3e354d06305e" />

Note that placing agents is not the part of the generator code. This is done by a separate scene in config (`Random`).

berekuk commented:
A bit more complex example: "dungeon with 20 rooms, one room is for treasure, 3 other rooms have 2 agents each, and 2 more rooms are mazes".

<img width="706" alt="Screenshot 2025-04-07 at 22 32 54" src="https://github.com/user-attachments/assets/85ff15bd-64d1-4905-af72-75e9a66821ee" />

Full config:

```
_target_: mettagrid.map.mapgen.MapGen

width: 100
height: 120
border_width: 1

root:
  _target_: mettagrid.map.scenes.bsp.BSP
  rooms: 20
  min_room_size: 3
  min_room_size_ratio: 0.5
  max_room_size_ratio: 0.9

  children:
    - limit: 1
      lock: lock1
      scene:
        _target_: mettagrid.map.scenes.random.Random
        objects:
          generator: 3
          altar: 3

    - limit: 3
      lock: lock1
      scene:
        _target_: mettagrid.map.scenes.random.Random
        agents: 2

    - limit: 2
      lock: lock1
      scene:
        _target_: mettagrid.map.scenes.maze.MazeKruskal
        extra_space_empty: true
```

PS: this example also shows the limits of composability - in this case I got lucky, but in other generations the maze entrances were often blocked. This could be solved by registering the corridor entrance points, and then passing them to maze generators as parameters somehow, but it's not trivial.

berekuk commented:
First take at porting https://github.com/mxgmn/ConvChain/ (same author as WFC, but slightly simpler - there's less code there port).

Four different configs here, see https://github.com/Metta-AI/mettagrid/pull/73/commits/239fd68c3976357e165081d4f4b959ed2d181562#diff-66496193cb00bdb19f3b134184caf19df60b85eff1386d9b4785d818a2c052c4 for details.

Claude 3.7 one-shotted most of the algorithm conversion from C#, tbh I haven't even read it carefully yet.

- no guarantees of connectivity yet (WFC would be more reliable here, ConvChain is more random)
- it's kind of slow, heavily depends on N in NxN pattern; O(2^(N*N)), so N=4 is not viable
- `iterations` and `temperature` heavily affect the output, when I set `iterations` to be too large it deteriorates into an empty field. So I think it's not possible to set these randomly and get a good result every time

<img width="740" alt="Screenshot 2025-04-08 at 20 11 32" src="https://github.com/user-attachments/assets/3351165f-3510-4ed8-8388-6dd295a36c7a" />

sasmith commented:
I'm currently reviewing just at a looking-at-the-conversation level, since I think you're looking for high-level directional feedback, not a code review at the moment. Lmk if I'm wrong about that.
* This looks good directionally.
* I think we want non-rectangular rooms, beyond just corridors (although they're a good example. As concrete examples, we should be able to tile with triangles or hexagons. That said, I expect rectangles to be the dominant room type, so having rectangles that can fill the space without needing to be cleanly aligned with each other (which I think some of your examples are?) seems more important than different shapes.
  * I think your "non rectangular areas" future focus is good here. Maybe generic convex polygons? Concave would be nice as well, but I can see a lot of reasons why concave could be easier.
* Symlinks seem like something we can figure out.
* I agree that we shouldn't be declaring num_agents twice. It's implicitly defined in the map, and it's probably valuable to have this denormalized (so you don't need to read the whole map to see how many agents there are), but it doesn't need to exist separately from the map.
  * It could also be helpful to think of this mapgen as something that should work standalone from mettagrid. I think this should impact how we think about the mapgen's config.
  * For most games (e.g., starcraft), you can play with fewer players than the maximum the map allows. I don't think we need to allow that here -- we can just generate a different map. (I don't think this matters very much for what you're doing, but wanted to share my mental model.)
  * Generally, I'd nudge you to focus primarily on map shape, and only secondarily think about content. I.e., focus on the layout of walls, versus agent / resource placement. Like, I think we'll want to be able to say "there should be an agent in one of these three rooms", but I think this sort of thing it something we can solve on top of map generation.
* Nitpick regarding the corridors: I don't think we want them to be long, and I think we want them to (usually) be wider. In particular, agents spending a lot of time walking through corridors is going to be boring, and we also don't want agents to block each other in corridors of they're going opposite directions.
* The ConvChain maps look great and very organic. I hear you saying that they're expensive. As a rule of thumb, I think we should aim for map generation to cost less than 2% percent of whatever we're spending on training. So map generation taking "a few seconds" seems good, and taking "a few minutes" seems bad. If you're getting great results and it's too slow, that still seems worth checking in, and just highlighting that it's something we'll need to invest in making faster.

berekuk commented:
> I think you're looking for high-level directional feedback, not a code review at the moment

High-level is good. I'd like to get a more specific feedback on one particular question: does the config approach in [mapgen_demo.yaml](https://github.com/Metta-AI/mettagrid/pull/73/files#diff-da942bba339bb0f26641d4e1965f86ed0442016573fab1e1922306680bf783ff) look good to you? If yes, and I hope so, then it might be worth it to convert the remaining maps to use it quickly and retire the old `mettagrid.config.room.*` system (this shouldn't take me too long, I've converted most of things).

That would be a clear checkpoint for this PR. Or we could merge this PR first and then I'd do the remaining conversion later.

> I think we want non-rectangular rooms, beyond just corridors
> Nitpick regarding the corridors: I don't think we want them to be long, and I think we want them to (usually) be wider

Yes, agreed. In BSP maps, room sizes can be increased (`min_room_size_ratio`, `max_room_size_ratio`), and I could tune the corridor widths. With corridors in BSP, the checks that corridors don't merge with rooms in ugly way are a bit tricky, but that's solvable, and might not even be important.

> I think we'll want to be able to say "there should be an agent in one of these three rooms", but I think this sort of thing it something we can solve on top of map generation.

There's a dimension to this project that we've discussed on Discord, about hands-off approach (where map is generated based only on `size` and `complexity` parameters) vs finer control (where the semantics are described in configs).

The initial version of this PR was more focused on finer control, and one of the key ideas here is that the output of a scene is not just a grid, but a set of "areas". After I've switched more to hands-off, and started looking into things like ConvChain and WaveFunctionCollapse, I became less sure about this — when the map is generated with ConvChain, the generator doesn't know anything about the map's semantics.

Still, I think it could be neat to have the support for both approaches in the same system, so that specific eval maps could be built with the same system. Also, in some cases we could extract features from the generated map after the fact.

I have two big directions I want to try next:
1. WaveFunctionCollapse with many patterns - maybe it'll generate enough complexity that a single WFC scene is enough for all interesting maps.
2. Fractal scenes that are composed of modules, but the hierarchy and children scenes is determined randomly instead of written out explicitly in the config. This will require a few more types of scenes, some of which are "maze creators", other are "object creators", and others are "embellishers" (transform the maze, drop obstacles in free spaces, randomize walls), etc.

How (2) might work: the top-level MapGen module is parameterized with width/height and some complexity budget.
It randomly picks the children structure, then a random scene for each child, and each random scene does the same. The complexity budget is divided by children and passed down until it's exhausted.

(2) is clearly more complicated than (1) and hand-crafting any single map generator. At the same time, it could be very powerful and flexible. Right now I'm focusing more on WFC, and will decide if we need something more after we see the diversity we can get from WFC and ConvChain approaches.

> So map generation taking "a few seconds" seems good, and taking "a few minutes" seems bad.

Got it. We can also make the generation phase offline, and pick the maps from the pre-generated database for training, I think.

berekuk commented:
> It's implicitly defined in the map, and it's probably valuable to have this denormalized (so you don't need to read the whole map to see how many agents there are), but it doesn't need to exist separately from the map. 

Counting agents in the map sound like something that can be done very easily though?

We might want to know agent count in advance if map loading is slow, if we want to load policies and generate maps in parallel. Not sure how useful this is.

There are two levels to this question btw:
1. map _config_ drives the agent count
2. generated map drives the agent count

I'm mostly interested in (1) right now; this is because I constantly have to do `python -m tests.show_map game/map_builder=mapgen_convchain  game.num_agents=4`.

I can't put the num_agents in `game/map_builder/mapgen_convchain.yaml`, because it's higher in the hierarchy. And I can't make use of `num_agents` in the map generator in the general case, the `num_rooms: ${div:${..num_agents},6}` trick from the older `simple.yaml` config doesn't generalize.

Of course, maybe I should just move configs to `game/`, and put this there...
```
game:
  num_agents: ...
  map_builder: ...
```

At the same time, I think doing (2), i.e. counting agents generated by the builder after the fact, would be fine, and cleaner than the current approach. If you agree, I can send this as a separate smaller PR.

sasmith commented:
I don't completely follow what you're proposing, so I'll say how I'm thinking about things. I also don't think I have a model of "the one right way to do things", so I'm happy for your input on what you think would be more successful.

One-step process
There's a `map generator config` that gives input the the map generator. Amongst other things, this config says how many agents should be on the generated map. mettagrid never needs to look at the `map generator config`, it only looks at the map; so from that perspective, the number of agents is "set by the map".

Two-step process
There's an initial map generation stage that generates a map with no information about agents (or other artifacts on the map) -- it just generates terrain. There's a second stage that places artifacts on the map, and the configuration for this second stage includes a number of agents to place. Once again, mettagrid doesn't need to read the config -- it just reads the number of agents from the config.


However, in both cases, I would expect there's some higher-level process that's determining how many agents there should be, and what models should be used, and things like this. And I'd expect this higher level process to be controlling which maps are used in a way that controls how many agents should be placed on the map. E.g., it could be writing the second-stage map generation configs. I don't _think_ this matters from the point of view of map generation; but it could lead to some duplication of how we specify agents. E.g., there could reasonably be a part of the config that say "we want to train agents in a 1v1 setting, so we should generate maps with 2 agents", and then if the map generation generates maps with 3 agents, we'll hit some sort of exception, since we won't be able to correctly place the two agents as planned.

> At the same time, I think doing (2), i.e. counting agents generated by the builder after the fact, would be fine, and cleaner than the current approach. If you agree, I can send this as a separate smaller PR.

If this is straightforward, then yes, send me the PR. I think that'll clarify some confusion I have.

====

Overall, though, if we can think of this as a two step process, where we're first generating terrain and secondarily adding agents and other artifacts, then we can delay having to think about num agents. Would that help?

berekuk commented:
I think one-step or two-steps (or more! the new architecture in this PR is largely about an arbitrary number of steps) as mostly independent from the num_agents change.

Might be easier to give an example. This config produces 8 agents:

```
_target_: mettagrid.map.mapgen.MapGen

width: 50
height: 40

root:
  _target_: mettagrid.map.scenes.room_grid.RoomGrid
  rows: 2
  columns: 3
  children:
  # Add cylinders in two random rooms
  - limit: 2
    scene:
      _target_: mettagrid.map.scenes.inline_ascii.InlineAscii
      data: |
        WWWWWWWWWW
             A    
        WWWWWWWWWW
  # Add one agent in each room
  - scene:
      _target_: mettagrid.map.scenes.random.Random
      agents: 1
```

The number of agents is static and driven by the config. The config can only be interpreted in two ways:
1. By a user who reads the config and knows now scenes work
2. By producing the map and then counting the agents

It's possible to make scenes that would add a random number of agents, but that depends on the specific configuration. It's possible to write a configuration that makes the number of agents completely predictable (and still driven by MapGen).

Passing the agents in a top-down fashion in this approach won't work well in general, though: I'm aiming to make MapGen flexible to make the decisions on what to put in the map on each level of scene tree. This makes the system expressive while keeping the control over the process.

The question about num_agents for me is about whether we can always grab their number from the environment. My impression is that we can, and `num_agents` mostly serves as an assertion. (We can keep it as an assertion, even, but make it optional.)

I also want to highlight that this is **not a blocker**. This PR works with explicit `num_agents`, there's just a minor usability issue because of duplication. After looking at all places in the codebase that refer to `num_agents`, I think it's better if I do this after this PR is merged (some parts are tricky).

berekuk commented:
I've marked this PR as ready for review.

I could continue to build upon it, but it seems useful to get this in the main so that others could run trains with new maps.

There's one blocker here: [Mettagrid symlinks refactor](https://github.com/Metta-AI/metta/pull/124) PR **must be merged before this**. After that, there are no breaking changes here.

berekuk commented:
More scenes and features:

WFC is now much faster. 400x400 map takes 32 seconds (on M1 Pro), performance scales linearly. This is ~16x slower than most other generators like maze. I don't think I can optimize it significantly more without rewriting in C++.

Maze scene allows arbitrary values for room_size and wall_size:
<img width="592" alt="Screenshot 2025-04-13 at 22 02 26" src="https://github.com/user-attachments/assets/4e3582a2-d53f-4655-b8dd-f58fda854bed" />

New MakeConnected scene that can be applied on top of any output to make it connected; example for 4 rooms grid -> convchain -> make_connected:
<img width="817" alt="Screenshot 2025-04-13 at 21 58 11" src="https://github.com/user-attachments/assets/5dc78ad5-8b49-4d3a-8a21-5a37cac234d0" />

The MakeConnected algorithm sometimes it drills too many tunnels because I create tunnels from the largest connected component to all others, and doesn't check for duplicate work; finding the minimal subset of walls to remove is not trivial. But it does always succeed and it's guaranteed that it won't explode in complexity. Not sure if spending more time on it is the high priority.

Mirror scene that creates a sub-scene and reflects it horizontally, vertically or both.
Here's an example with 4 rooms for the same WFC pattern. First one is unmirrored, others show various mirror modes:
<img width="598" alt="Screenshot 2025-04-13 at 22 12 33" src="https://github.com/user-attachments/assets/cfd1e6b5-3727-4e8e-8b6c-efe3cee3fdfd" />

berekuk commented on mettagrid/map/node.py:
```diff
@@ -0,0 +1,98 @@
+import numpy.typing as npt
+
+
+class Area(TypedDict):
```
Oh right, that's more idiomatic.
(It's been a few years since the last time last programmed in Python, this is accidental TypeScript-ism, an attempt to define a type only and keep using dicts because those feel more lightweight. But that's not really true in Python.)

berekuk commented on configs/game/map_builder/mapgen_demo.yaml:
```diff
@@ -0,0 +1,103 @@
+      _target_: mettagrid.map.scenes.random.Random
+
+      # Let's place some objects in each room.
+      # Note that for the maze rooms, Random scene will take existing walls into account.
```
Oh. Maybe I should've looked into existing object behaviors more, I mostly ignored the specifics and assumed that they're all swappable.

I think the right thing to do here is to make `MakeConnected` scene configurable, pass in object types that should be treated like walls. It's somewhat tricky though, because `MakeConnected` creates tunnels by turning walls into empty objects. If we sprinkle on objects and agents first, then dig, then a random number of agents could be deleted.

So the correct pattern (for generating a random well-formed map) is something like:
1. ...make any walls-only map
2. Random (objects)
3. MakeConnected (dig through walls and mines)
4. Random (agents)

sasmith replied:
The objects I'm talking about are things like mines and altars, which are permanent fixture (maybe can be destroyed based on map configurations, but not moved). We have recently introduced "blocks" which are movable walls, so we have that complication too.

I think there are a few options, in addition to what you suggested:
  * Replace walls with the objects. For this, only select walls that are bordered by at least one empty space.
    * This might be fine. It also might and up with crowded objects.
  * Add objects in spaces that are already "open". E.g., if the space is surrounded on all eight sides by empty squares, then the object won't be blocking any paths.
    * I somewhat prefer this. I think it'll make the most "natural" maps.
    * It requires that rooms be sufficiently large to hold the needed objects, but this seems more natural than needing to carve out walls in order to fit things.

berekuk replied:
> Add objects in spaces that are already "open". E.g., if the space is surrounded on all eight sides by empty squares, then the object won't be blocking any paths. 

Don't we want agents to learn to interact with objects in corners though, and all other possible situations? (Example: cylinder cognitive_eval which intentionally places the altar in a narrow corridor)

berekuk commented on configs/game/map_builder/mapgen_demo.yaml:
```diff
@@ -0,0 +1,103 @@
+  #
+  # Lock will exclude the areas that were already selected by previous children that use the same lock name.
+  # So this guarantees that the custom map will not overlap with the maze.
+  - lock: lockname
```
Yeah, I've considered multiple locks but couldn't think of any real-world examples where they'd be worth it yet. Seems easy to add later.

Very open to alternatives to "lock", couldn't find any.

berekuk commented on configs/game/map_builder/mapgen_demo.yaml:
```diff
@@ -0,0 +1,103 @@
+width: 50
+height: 40
+
+# Outer border width. Defaults to 6.
```
Yep, this is only for the root.

Lack of layout capabilities (where child scenes affect the size of child scenes) is the important limitation of this approach, so scenes don't have "width" and "height" at all.

Btw, I've updated the default width to 1 in the latest commits.

berekuk commented on mettagrid/map/mapgen.py:
```diff
@@ -0,0 +1,48 @@
+        self._grid[:, :border_width] = "wall"
+        self._grid[:, -border_width:] = "wall"
+
+        # TODO - add outer wall borders, always useful
```
I've removed the comment, it was stale.

I don't think it's useful to have borders for most scenes, since we want them to be connected (with the exception of RoomGrid).

berekuk commented on mettagrid/map/node.py:
```diff
@@ -0,0 +1,98 @@
+        lock = query.get("lock")
+        if lock:
+            if lock not in self._locks:
+                self._locks[lock] = []
```
changed, thanks

sasmith commented on mettagrid/map/scenes/make_connected.py:
```diff
@@ -0,0 +1,167 @@
+
+    It does this by:
+    - Finding all the connected components
+    - Digging shortest tunnels from the largest component to all other components
```
Consider if we want to join the two closest components instead? I guess that would make it n**2, but it seems like another natural way to join things.

Edit: I understand this better now, and think this might be something to go into the TODO that I'm proposing.

sasmith commented on mettagrid/map/scenes/make_connected.py:
```diff
@@ -0,0 +1,167 @@
+    - Finding all the connected components
+    - Digging shortest tunnels from the largest component to all other components
+
+    TODO: This can result in some extra tunnels being dug.
```
Surprising!

But also, I think for most maps I'd want some extra tunnels dug. It seems like the intended algorithm would result in a minimally connected graphs (with regards to inter-node connections). I think sometimes we'd want that, and sometimes we'd want more than minimally connected. Consider adding some option for extra edges, which we'd add by picking two random nodes and connecting them, or something.

sasmith commented on mettagrid/map/scenes/make_connected.py:
```diff
@@ -0,0 +1,167 @@
+from mettagrid.map.node import Node
+from mettagrid.map.scene import Scene
+
+directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
```
prefer `DIRECTIONS`, to make it easier to identify as some sort of constant

sasmith commented on mettagrid/map/scenes/make_connected.py:
```diff
@@ -0,0 +1,167 @@
+        distances = np.full((height, width), np.inf)
+        queue = []
+        for cell in component_cells:
+            distances[cell[0], cell[1]] = 0
```
If you want to be cute, consider `distances[*cell] = 0`

sasmith commented on mettagrid/map/scenes/make_connected.py:
```diff
@@ -0,0 +1,167 @@
+        logger.debug(f"Largest component: {largest_component_id}")
+
+        logger.debug("Populating distance to largest component")
+        distances_to_largest_component = self._distance_to_component(
```
Ah! I expected this to work recursively, where once we connected something, we would then consider the newly connected component to be part of the largest component, and figure out the next connection based on this (so, e.g., the potential set of connection points would be increased). But we don't. Instead we're going to keep connecting things to the original largest component.

This seems fine, but I think I'd want to TODO about iterating on this. I see that performance could get really bad here if we need to repeatedly crawl the whole grid, but I expect we could make this more efficient by doing things like only considering exterior points and such.

Also, I think we should be more explicit about this being the logic (that everything connects to the originally largest component, vs it being iterative).

berekuk replied:
Yes. LLMs suggested minimal spanning tree, but then naively I'd need to precalculate the minimal distance between each pair of components; I couldn't figure out how to make it definitely fast on the spot (component_count * map_size is still slow, what if we have a lot of components; I can't make any assumptions about the underlying structure, it could be lots of isolated 1x1 cells in the worst case, or weird concave shapes). Seemed solvable, but this was good enough for my goals, for now, didn't want to complicate things prematurely.

At the same time, it's annoyingly not good enough for BSP, or for mazes. (theoretically we could do RoomGrid of 1x1 rooms -> MakeConnected and get a maze... but not with this approach; this approach destroys most of the map).

I'll add a todo note, and spend some more time thinking if there's some easy way to do this that I've missed.

sasmith commented on mettagrid/map/scenes/make_connected.py:
```diff
@@ -0,0 +1,167 @@
+            # find the cell that's closest to the largest component
+            min_distance = np.inf
+            min_distance_cell: Union[Tuple[int, int], None] = None
+            for cell in component_cells[component_id]:
```
min_distance_cell = min(cells, key: lambda c: distances_to_largest_component[*c])

sasmith commented on mettagrid/map/scenes/make_connected.py:
```diff
@@ -0,0 +1,167 @@
+                # Find the neighbor with the smallest distance to the largest component
+                y, x = current_cell
+                # Find all neighbors with the smallest distance to the largest component
+                candidates = []
```
With every step, the distance should go down by 1. So we can probably clean this up a bit by having the candidates be all cells that are 1 closer than the current distance.

sasmith commented on mettagrid/map/scenes/make_connected.py:
```diff
@@ -0,0 +1,167 @@
+                    y, x = queue[i]
+                    i += 1
+                    if components[y, x] != -1:
+                        continue
```
Could have an assertion that `components[y, x] == component_id`.

berekuk replied:
I've rewritten this as `visited` (with bools), I never actually used indices.

Btw, I've just fixed a dumb bug here: `y` and `x` were overwriting outer `y` and `x`, so some rooms were not connected. (implicit variable declarations might be my least favorite Python feature...)

I've spotted this on large maps by accident; added an assertion at the end just to be safe: `assert len(self._make_components(node)) == 1`.

sasmith commented on mettagrid/map/scenes/make_connected.py:
```diff
@@ -0,0 +1,167 @@
+                components_cells.append([])
+                queue = [(y, x)]
+                i = 0
+                while i < len(queue):
```
This is fine as is, but you could consider using a stack here, and not needing `i`.

sasmith commented on mettagrid/map/scenes/make_connected.py:
```diff
@@ -0,0 +1,167 @@
+            queue.append(cell)
+
+        i = 0
+        while i < len(queue):
```
Same as above, I think this could be a stack and then not need `i`. Or if you really wanted to make the point that order doesn't matter, it could be a set. So

```
to_visit = set(...)
while to_visit:
  y, x = to_visit.pop()
  ...
```

berekuk commented on mettagrid/map/scenes/convchain.py:
```diff
@@ -0,0 +1,78 @@
+        self._weights = ascii_to_weights_of_all_patterns(
+            self._pattern, self._receptor_size
+        )
+        # Ensure all weights are positive
```
This is justified in (3) here: https://github.com/mxgmn/ConvChain/

> For the speed of convergence, it's convenient for weights of all patterns to be nonzero. So let's redefine the weight(P) of a pattern P to be the number of patterns P in the input if that number is more than zero and some small number eps otherwise, 0 < eps < 1.

But also tbh I don't really understand how ConvChain works yet... I've converted it from C# with Claude and verified that it matches the algorithm, but didn't have time to figure out what's going on, because I tried to cover as much as possible in a few days, so I've switched to trying WFC instead (which I do understand).

I'll get back to it soon.

sasmith commented on mettagrid/map/scenes/wfc.py:
```diff
@@ -0,0 +1,293 @@
+
+        self._fill_propagator()
+
+    def _fill_propagator(self):
```
"Here we pre-calculate pattern compatibility, and we'll use this to check the compatibility of potential patterns going forward. An alternative could be to write (/ flatten) chosen patterns to the grid, and then check potential patterns against the grid; but this is easier."

berekuk replied:
> An alternative could be to write (/ flatten) chosen patterns to the grid, and then check potential patterns against the grid; but this is easier.

I'm not sure what you mean by this:
- the main purpose of the propagator is performance optimization
- I don't see how we could write patterns to the grid without it; the wave function is expressed in pattern ids. these could be decoded to patterns and compared on each step, but there's no grid to write them on (since we maintain the list of possibilities that's narrowed incrementally)

Rephrased so:
> Here we pre-calculate pattern compatibility, and we'll use this to check the compatibility of potential patterns going forward, when we exclude some possibility from the wave function and want to propagate the change.

PS: General thought on comments, for this particular case: it's hard to guess the level of competence of people who would read this code later. The range of algorithmic competence among engineers is wide, from "this is obvious" to "ugh I beat my head against this code for a few days before I got it" (pre-LLMs and pre-Codeforces me would be in this latter category). LLMs might have an advantage here, because you can ask it specific questions.

sasmith reviewed (changes requested):
Some additional feedback. Nothing that should be blocking, but I'm marking as requesting changes due to the number of comments, and my inability to keep the whole diff in my head at once.

berekuk commented on mettagrid/map/scenes/bsp.py:
```diff
@@ -0,0 +1,352 @@
+        zones = [Zone(0, 0, grid.shape[1], grid.shape[0])]
+
+        for _ in range(self._rooms - 1):  # split rooms-1 times
+            zone = zones[next_split_id]
```
I needed non-leaf zones too, for corridors, later in the code.

I've refactored this code  - there's a `BSPTree` class now that encapsulates most of the logic, and I'm reusing it in the new `BSPLayout` scene that's used in `Auto` scene that I got in after your review.

berekuk commented on mettagrid/map/scenes/bsp.py:
```diff
@@ -0,0 +1,352 @@
+    def from_zone(
+        grid: np.ndarray, zone: Zone, side: Literal["up", "down"]
+    ) -> "Surface":
+        # Scan the entire zone, starting from the top or bottom, and collect all the y values that are part of the surface.
```
The zone is rectangular, but the shape inside the zone can be pretty complex.

The reason why I needed all this is pretty technical and I'm not sure it's worth improving it. There might be an easier way, or maybe it's extra complexity that could be ignored.

Consider this maze:
```
################
#             ##
#    #####    ##
#    #####    ##
################
################
################
####       #####
####       #####
################
```

There are three rooms here, two on the top that are already connected, and one on the bottom.

[One of the guides that I followed](https://www.roguebasin.com/index.php/Basic_BSP_Dungeon_generation) said "we can connect two sub-regions with a link either between two rooms, or a corridor and a room or two corridors."

So I needed to connect the bottom room with either the left room, right room or the corridor. And the top structure could be more complicated than this example (the corridor could have turns, there could be more rooms and more corridors, etc.)

The most annoying special case is that corridors can sometimes graze against the space of an adjacent room, like this:
```
################
#             ##
#    C####    ##
#    C####    ##
#####C##########
#####C##########
#####C##########
####       #####
####       #####
################
```

Ideally, we want to avoid that. The surface code is supposed to do this, but I'm not sure it handles all cases perfectly. I could do something more clever like pathfinding, not sure if it's worth extra effort though.

berekuk commented on mettagrid/map/scenes/bsp.py:
```diff
@@ -0,0 +1,352 @@
+        self.start = start
+        self.length = length
+
+    def transpose(self) -> "Line":
```
I've added some comments.

berekuk commented on mettagrid/map/scenes/wfc.py:
Yeah... 4d arrays for propagator and compatible are particularly hard to explain even in comments, I don't see a great way for people to grok this without careful reading through both the documentation and the source (which is what I did with the original C#, with some help from o1-pro to check that I haven't messed up the indices somewhere).

If I get back to this code later, e.g. to make it more flexible for some special case that we need, I'll try to add more comments and tests.

berekuk commented on mettagrid/map/utils/pattern.py:
```diff
@@ -0,0 +1,184 @@
+            p7 = p6.reflected()
+            return [p0, p1, p2, p3, p4, p5, p6, p7]
+        elif symmetry == "horizontal":
+            return [self, self.reflected()]
```
I ported these from WFC and ConvChain, mostly (original is [here](https://github.com/mxgmn/WaveFunctionCollapse/blob/a088882ea7261ca11f7e02b58b12a870d809e8a4/OverlappingModel.cs#L53) if you're curious).

The reasoning is:
- I wasn't sure what I would need
- I wanted to try the examples from WFC repo to see if my outputs are similar (still haven't done that in a systematic way; you can see their examples [here](https://github.com/mxgmn/WaveFunctionCollapse/blob/master/samples.xml); symmetry=2 is horizontal)
- symmetry="all" is definitely useful in some cases; and "horizontal" is a subset of that, so it was trivial to add both

berekuk commented:
I've responded to everything in your comments; please say if you think I pushed back on something you cared about.
 
Added some more new code here, mostly related to Auto scenes.

Remerged with fresh main.

There's still more work to be done, but generally I think this is in good shape, and could be used as is for training with the latest Auto improvements.

berekuk commented on configs/game/map_builder/mapgen_auto.yaml:
```diff
@@ -0,0 +1,173 @@
+      mine:
+        distribution_type: uniform
+        min: 0.001
+        max: 0.03
```
Sorry for the inconsistency here vs `min_*` fields in other parts of this config! I started with naive min/max, and then implemented these more generic "distribution_types" for objects in rooms specifically, because I wanted to clearly see that distributions are clearly different, and uniform was not diverse enough.

I'd like to use these distribution type objects everywhere, eventually, but want to get the confirmation first that the pattern looks ok (semi-expect someone to push back in the direction of resolvers because of the need for sweeps).

berekuk commented on mettagrid/map/utils/random.py:
```diff
@@ -0,0 +1,52 @@
+    log_high = np.log(high)
+
+    # Calculate normalized sigmas using the inverse of the normal CDF
+    normalized_sigmas = stats.norm.ppf(1 - (1 - probability) / 2)
```
The original code I linked above was well-tested, so it should be fine.
